src/ui/public/saved_objects/saved_objects_client.js

```diff
@@ -0,0 +1,104 @@
+import { resolve as resolveUrl, format as formatUrl } from 'url';
+import { pick, partial } from 'lodash';
+import { keysToSnakeCaseShallow, keysToCamelCaseShallow } from '../../../utils/case_conversion';
+
+import { SavedObject } from './saved_object';
+
+const join = (...uriComponents) => (
+  uriComponents.filter(Boolean).map(encodeURIComponent).join('/')
+);
+
+export class SavedObjectsClient {
+  constructor($http, basePath, PromiseCtor = Promise) {
+    this._$http = $http;
+    this._apiBaseUrl = `${basePath}/api/saved_objects/`;
+    this._PromiseCtor = PromiseCtor;
+  }
+
+  get(type, id) {
+    if (!type || !id) {
+      return this._PromiseCtor.reject(new Error('requires type and id'));
+    }
+
+    return this._request('GET', this._getUrl([type, id])).then(resp => {
+      return new this.ObjectClass(resp);
+    });
+  }
+
+  delete(type, id) {
+    if (!type || !id) {
+      return this._PromiseCtor.reject(new Error('requires type and id'));
+    }
+
+    return this._request('DELETE', this._getUrl([type, id]));
+  }
+
+  update(type, id, attributes, { version } = {}) {
+    if (!type || !id || !attributes) {
+      return this._PromiseCtor.reject(new Error('requires type, id and attributes'));
+    }
+
+    const body = {
+      attributes,
+      version
+    };
+
+    return this._request('PUT', this._getUrl([type, id]), body);
+  }
+
+  create(type, body) {
+    if (!type || !body) {
+      return this._PromiseCtor.reject(new Error('requires type and body'));
+    }
+
+    const url = this._getUrl([type]);
+
+    return this._request('POST', url, body);
+  }
+
+  find(options = {}) {
+    const url = this._getUrl([options.type]);
+
+    return this._request('GET', url, keysToSnakeCaseShallow(options)).then(resp => {
+      resp.data = resp.data.map(d => new this.ObjectClass(d));
+      return keysToCamelCaseShallow(resp);
+    });
+  }
+
+  get ObjectClass() {
```

Comment:
Using a getter here returns a new constructor every time `ObjectClass` is accessed. Besides being somewhat wasteful of resources it also gives the illusion that

```js
savedObjectInstance instanceof savedObjectsClient.ObjectClass
```

will evaluate to true. IMO, we should just have something like this:

```js
class SavedObjectsClient {
  // ...
  createSavedObject(options) {
    return new SavedObject(this, options)
  }
}
```